### PR TITLE
Select customer by Id

### PR DESCRIPTION
**Pull request by mamazu**
**Before:** You have to select an email address in the input field and a channel. Then with this email address a customer gets selected. If no customer is found one is created. Then in the rest of the process the email address is carried in the URL to identify the customer.

**Problems:** The URL is getting longer with a longer email address (the email address is in there twice at some time because a form submits it via post data) and the URL contains an `@` Symbol which could cause problems.

**Solution:** Select / Create and select the customer once the customer has been identified you get the id of the customer and use this in the URL.

**Changes:**
* The two forms for selecting and creating a customer now route to different end-points
* If you try to select a customer with an email that does not exist it throws an error

### CODE REVIEW

File: spec/Factory/OrderFactorySpec.php [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 namespace spec\Sylius\AdminOrderCreationPlugin\Factory;
 
+use InvalidArgumentException;
```
Comment (Zales0123) [inline spec/Factory/OrderFactorySpec.php]: Unneeded, we can just use `\InvalidArgumentException` in the code

File: spec/Factory/OrderFactorySpec.php [hunk 2]
```diff
@@ -71,40 +72,71 @@ function it_creates_order_for_customer_with_default_channel_locale_and_currency(
...
+    function it_throws_an_exception_if_the_customer_does_not_exist(
+        FactoryInterface $baseOrderFactory,
+        CustomerRepositoryInterface $customerRepository
+    ): void {
+        $customerRepository->find('1')->shouldBeCalled()->willReturn(null);
+
+        $baseOrderFactory->createNew()->shouldNotBeCalled();
+
+        $this->shouldThrow(InvalidArgumentException::class)
+            ->during('createForCustomerAndChannel', ['1', 'WEB-US']);
```
Comment (Zales0123) [inline spec/Factory/OrderFactorySpec.php]: Bad formatting:

```php
$this
    ->shouldThrow(\InvalidArgumentException::class)
    ->during('createForCustomerAndChannel', ['1', 'WEB-US'])
;
```

File: spec/Factory/OrderFactorySpec.php [hunk 3]
```diff
@@ -71,40 +72,71 @@ function it_creates_order_for_customer_with_default_channel_locale_and_currency(
...
-        $customerRepository->findOneBy(['email' => 'customer@example.com'])->willReturn(null);
 
-        $customerFactory->createNew()->willReturn($customer);
-        $customer->setEmail('customer@example.com')->shouldBeCalled();
+        $customerRepository->find('1')->willReturn($customer);
+        $channelRepository->findOneByCode('WEB-US')->willReturn($channel);
+
+        $channel->getBaseCurrency()->willReturn(null);
+
+
```
Comment (Zales0123) [inline spec/Factory/OrderFactorySpec.php]: Doubled blank line

File: src/Controller/CustomerCreationAction.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        $customer = $this->customerProvider->provideNewCustomer($request->attributes->get('customerEmail'));
+
+        return new RedirectResponse(
+            $this->router->generate(
+                'sylius_admin_order_creation_order_create', [
+                'customerId' => $customer->getId(),
+                'channelCode' => $request->attributes->get('channelCode'),
+            ]
+            )
+        );
```
Comment (Zales0123) [inline src/Controller/CustomerCreationAction.php]:
```php
return new RedirectResponse(
    $this->router->generate(
        'sylius_admin_order_creation_order_create', [
            'customerId' => $customer->getId(),
            'channelCode' => $request->attributes->get('channelCode'),
        ]
    )
;
```

File: src/Factory/OrderFactoryInterface.php [hunk 5]
```diff
@@ -9,7 +9,18 @@
 
 interface OrderFactoryInterface extends FactoryInterface
 {
-    public function createForCustomerAndChannel(string $customerEmail, string $channelCode): OrderInterface;
+    /**
+     * Creates a new order for a given customer in a given channel.
+     * 
+     * @param string $customerId
+     * @param string $channelCode
+     */
```
Comment (Zales0123) [inline src/Factory/OrderFactoryInterface.php]: The whole docblock is redundant, the same below

File: src/Provider/CustomerProvider.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+final class CustomerProvider implements CustomerProviderInterface
+{
+    /** @var CustomerRepository */
+    private $customerRepository;
+
+    /** @var FactoryInterface */
+    private $customerFactory;
+
+    public function __construct(
+        CustomerRepository $customerRepository,
```
Comment (Zales0123) [inline src/Provider/CustomerProvider.php]: Should be `CustomerRepositoryInterface`

File: src/Provider/CustomerProvider.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    ) {
+        $this->customerRepository = $customerRepository;
+        $this->customerFactory = $customerFactory;
+    }
+
+    /** {@inheritdoc}  */
+    public function provideExistingCustomer(string $id): CustomerInterface
+    {
+        $customer = $this->customerRepository->find($id);
+        assert($customer instanceof CustomerInterface);
```
Comment (Zales0123) [inline src/Provider/CustomerProvider.php]: As we have a proper extension to PHPStan we can use `Webmozart/Assert`

File: src/Provider/CustomerProvider.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    /** {@inheritdoc}  */
+    public function provideNewCustomer(string $email): CustomerInterface
+    {
+        $customer = $this->customerFactory->createNew();
+        assert($customer instanceof CustomerInterface);
+
+        $customer->setEmail($email);
+
+        assert($customer instanceof CustomerInterface);
```
Comment (Zales0123) [inline src/Provider/CustomerProvider.php]: Why checking it twice?

File: src/Provider/CustomerProvider.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\AdminOrderCreationPlugin\Provider;
+
+use Sylius\AdminOrderCreationPlugin\Doctrine\ORM\CustomerRepository;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+
+final class CustomerProvider implements CustomerProviderInterface
```
Comment (Zales0123) [inline src/Provider/CustomerProvider.php]: There is no spec test for this class

File: src/Provider/CustomerProviderInterface.php [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+interface CustomerProviderInterface
+{
+    /**
+     * Loads an existing customer from the database and throws an error if it does not exist
+     *
+     * @param string $id
+     *
+     * @return CustomerInterface
+     */
```
Comment (Zales0123) [inline src/Provider/CustomerProviderInterface.php]: Same here, redundant docblocks

File: src/Provider/CustomerProviderInterface.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    public function provideExistingCustomer(string $id): CustomerInterface;
```
Comment (Zales0123) [inline src/Provider/CustomerProviderInterface.php]: And why `$id` is `string`? It should be `int`, I suppose

File: src/Resources/translations/messages.en.yml [hunk 12]
```diff
@@ -17,3 +17,4 @@ sylius_admin_order_creation:
         order_preview: 'Order preview'
         reorder: 'Reorder'
         shipping_methods_selection_requirement: 'You need to add some items and shipping address to select from eligible shipping method'
+        no_customer_selected: "You don't have a customer selected"
```
Comment (Zales0123) [inline src/Resources/translations/messages.en.yml]: Alphabetical order should be kept

File: src/Provider/CustomerProvider.php (hunk 8 above)
Comment (mamazu) [reply src/Provider/CustomerProvider.php]: Must have been a merge.

File: src/Provider/CustomerProviderInterface.php (hunk 11 above)
Comment (mamazu) [reply src/Provider/CustomerProviderInterface.php]: The idea behind this was that the Ids can also be UUIDs and therefore strings.

File: src/Provider/CustomerProvider.php (hunk 9 above)
Comment (mamazu) [reply src/Provider/CustomerProvider.php]: Should be done.

File: src/Resources/translations/messages.en.yml [hunk 13]
```diff
@@ -13,6 +13,7 @@ sylius_admin_order_creation:
         customer_email: 'Customer email cannot be empty'
         item_discount: 'Item discount'
         new_customer_email: 'New customer email'
+        no_customer_selected: "You haven't selected a customer"
```
Comment (bartoszpietrzak1994) [inline src/Resources/translations/messages.en.yml]: The message should be placed in single quotes.

Comment (mamazu) [reply src/Resources/translations/messages.en.yml]: But it contains a single quote. This would require escaping it and make it less readable.

Comment (bartoszpietrzak1994) [reply src/Resources/translations/messages.en.yml]: Yup, my bad!

**Comment by bartoszpietrzak1994** [comment]
The code seems ok for me. However, there's still one thing left regarding customerId requirements in routings.

After latest fixes you've removed the `+d` requirement only in some cases - what about sticking to the same requirement that was applied to `customerEmail`, which would be `customerId: .+` ?

**Comment by mamazu** [comment]
Isn't that implicitly done anyways?

**Comment by mamazu** [comment]
@bartoszpietrzak1994 Is there anything else?

**Comment by bartoszpietrzak1994** [comment]
@mamazu It's ok for me :)

**Comment by mamazu** [comment]
I just rebased it on the current master. So it's ready to merge.

**Pull request merged**